server/loginserver.ts

```diff
@@ -38,69 +36,67 @@ function parseJSON(/** @type {string} */ json) {
 	return data;
 }
 
-/** @typedef {[AnyObject?, number, Error?]} LoginServerResponse */
+type LoginServerResponse = [AnyObject | null, number, Error | null];
 
 class LoginServerInstance {
+	uri: string;
+	requestQueue: [AnyObject, (val: LoginServerResponse) => void][];
+	requestTimer: null;
+	requestLog: string;
+	lastRequest: number;
+	openRequests: number;
+	disabled: false;
+
 	constructor() {
 		this.uri = Config.loginserver;
-		/**
-		 * @type {[AnyObject, (val: LoginServerResponse) => void][]}
-		 */
 		this.requestQueue = [];
-
 		this.requestTimer = null;
-		/** @type {string} */
 		this.requestLog = '';
 		this.lastRequest = 0;
 		this.openRequests = 0;
 		this.disabled = false;
 	}
 
-	/**
-	 * @param {string} action
-	 * @param {AnyObject?} data
-	 * @return {Promise<LoginServerResponse>}
-	 */
-	instantRequest(action, data = null) {
+	instantRequest(action: string, data: AnyObject | null = null): Promise<LoginServerResponse> {
 		if (this.openRequests > 5) {
-			return Promise.resolve(/** @type {LoginServerResponse} */ (
+			return Promise.resolve((
 				[null, 0, new RangeError("Request overflow")]
 			));
 		}
 		this.openRequests++;
 		let dataString = '';
 		if (data) {
-			for (let i in data) {
+			for (const i in data) {
 				dataString += '&' + i + '=' + encodeURIComponent('' + data[i]);
 			}
 		}
-		const urlObject = url.parse(this.uri + 'action.php?act=' + action + '&serverid=' + Config.serverid + '&servertoken=' + encodeURIComponent(Config.servertoken) + '&nocache=' + new Date().getTime() + dataString);
+		const urlObject = url.parse(`${this.uri}action.php?act=action&serverid='${Config.serverid}&servertoken=` +
+		`'${encodeURIComponent(Config.servertoken)}&nocache=${new Date().getTime() + dataString}`);
 		return new Promise((resolve, reject) => {
-			// @ts-ignore TypeScript bug: http.get signature
-			let req = http.get(urlObject, res => {
-				Streams.readAll(res).then(buffer => {
-					let data = parseJSON(buffer).json || null;
-					resolve([data, res.statusCode || 0, null]);
+
+			const req = http.get(urlObject, (res: any) => {
```

Comment:
I'm not sure what type I should give `res` to.

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/https.d.ts and https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/http.d.ts should help answer that (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/http.d.ts#L177 appears to be the answer).

However, `Streams.readAll` just needs `NodeJS.ReadableStream`, which `stream.Readable` is (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/stream.d.ts#L20), so if you want to avoid `any`, `NodeJS.ReadableStream` seems correct

Reply:
Hm, that gives an error because `NodeJS.ReadableStream` does not have `statusCode`, I guess I could write a one-liner type for those?

Reply:
Maybe like:

```typescript
interface IncomingMessage extends NodeJS.ReadableStream {
  statusCode: number;
}
```

Reply:
Oh right yeah, I forgot about interfaces, lol. Thank you.